enterprise/internal/campaigns/service_apply_campaign.go

```diff
@@ -122,231 +122,121 @@ func (s *Service) ApplyCampaign(ctx context.Context, opts ApplyCampaignOpts) (ca
-		c, ok := r.changesetsByRepoHeadRef[k]
+		spec, ok := changesetSpecsByID[m.ChangesetSpecID]
 		if !ok {
-			// No, we don't have a changeset on that branch in this repo.
-			// We're going to create one so the changeset reconciler picks it up,
-			// creates a commit and pushes it to the branch.
-			// Except, of course, if spec.Spec.Published is false, then it doesn't do anything.
-			c, err = r.createChangesetForSpec(ctx, repo, spec)
-			if err != nil {
-				return err
-			}
-		} else {
-			// But if we already have a changeset in the given repository with
-			// the given branch, we need to update it to have the new spec
-			// and possibly re-attach it to the campaign:
-			if err = r.updateChangesetToNewSpec(ctx, c, spec); err != nil {
-				return err
-			}
+			// This should never happen.
+			return errors.New("spec not found")
 		}
-		// In both cases we want to attach it to the campaign
-		attachedChangesets[c.ID] = true
-	}
-
-	// We went through all the new changeset specs and either created or
-	// updated a changeset.
-	// Their IDs are all the IDs of changesets that should be in the campaign:
-	r.campaign.ChangesetIDs = []int64{}
-	for changesetID := range attachedChangesets {
-		r.campaign.ChangesetIDs = append(r.campaign.ChangesetIDs, changesetID)
-	}
 
-	// But it's possible that changesets are now detached, like Changeset 3 in
-	// the example above.
-	// This we need to detach and close.
-	for _, c := range r.changesets {
-		if _, ok := attachedChangesets[c.ID]; ok {
-			continue
+		// If we don't have access to a repository, we return an error. Why not
+		// simply skip the repository? If we skip it, the user can't reapply
+		// the same campaign spec, since it's already applied and re-applying
+		// would require a new spec.
+		repo, ok := accessibleReposByID[m.RepoID]
+		if !ok {
+			return &db.RepoNotFoundErr{ID: m.RepoID}
 		}
 
-		// If we don't have access to a repository, we don't detach nor close the changeset.
-		_, ok := r.accessibleReposByID[c.RepoID]
-		if !ok {
-			continue
+		if err := checkRepoSupported(repo); err != nil {
+			return err
 		}
 
-		if c.CurrentSpecID != 0 && c.OwnedByCampaignID == r.campaign.ID {
-			// If we have a current spec ID and the changeset was created by
-			// _this_ campaign that means we should detach and close it.
-
-			// But only if it was created on the code host:
-			if c.Published() {
-				c.Closing = true
-				c.ReconcilerState = campaigns.ReconcilerStateQueued
-			} else {
-				// otherwise we simply delete it.
-				if err = r.tx.DeleteChangeset(ctx, c.ID); err != nil {
-					return err
-				}
-				continue
+		var changeset *campaigns.Changeset
+
+		if m.ChangesetID != 0 {
+			changeset, ok = changesetsByID[m.ChangesetID]
+			if !ok {
+				// This should never happen.
+				return errors.New("changeset not found")
+			}
+			if spec.Spec.IsImportingExisting() {
+				err = r.attachTrackingChangeset(ctx, changeset)
```

Comment:
Nitpick: it doesn't attach the changeset, that happens further down.

```suggestion
				err = r.trackChangeset(ctx, changeset)
```

Reply:
It says "attach the changeset to the campaign of the rewirer" so I think it's fine